tls/s2n_server_key_exchange.c

```diff
@@ -29,71 +29,69 @@
     struct s2n_stuffer *in = &conn->handshake.io;
-    struct s2n_blob ecdhparams = {0};
-    struct s2n_blob signature = {0};
-    uint16_t signature_length;
+    struct s2n_blob data_to_sign = {0};
 
-    /* Read server ECDH params and calculate their hash */
-    GUARD(s2n_ecc_read_ecc_params(&conn->secure.server_ecc_params, in, &ecdhparams));
+    /* Read and process the KEM data */
+    GUARD(kem_core->server_key_recv(conn, &data_to_sign));
 
+    /* Add common signature data */
     if (conn->actual_protocol_version == S2N_TLS12) {
         s2n_hash_algorithm hash_algorithm;
         s2n_signature_algorithm signature_algorithm;
         GUARD(s2n_get_signature_hash_pair_if_supported(in, &hash_algorithm, &signature_algorithm));
-
-        GUARD(s2n_hash_init(&conn->secure.signature_hash, hash_algorithm));
+        GUARD(s2n_hash_init(signature_hash, hash_algorithm));
     } else {
-        GUARD(s2n_hash_init(&conn->secure.signature_hash, conn->secure.conn_hash_alg));
+        GUARD(s2n_hash_init(signature_hash, conn->secure.conn_hash_alg));
     }
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
-    GUARD(s2n_hash_update(&conn->secure.signature_hash, ecdhparams.data, ecdhparams.size));
+    GUARD(s2n_hash_update(signature_hash, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
+    GUARD(s2n_hash_update(signature_hash, conn->secure.server_random, S2N_TLS_RANDOM_DATA_LEN));
+
+    /* Add KEM specific data */
+    GUARD(s2n_hash_update(signature_hash, data_to_sign.data, data_to_sign.size));
 
     /* Verify the signature */
+    struct s2n_blob signature = {0};
+    uint16_t signature_length;
     GUARD(s2n_stuffer_read_uint16(in, &signature_length));
     signature.size = signature_length;
     signature.data = s2n_stuffer_raw_read(in, signature.size);
     notnull_check(signature.data);
     gt_check(signature_length, 0);
 
-    S2N_ERROR_IF(s2n_pkey_verify(&conn->secure.server_public_key, &conn->secure.signature_hash, &signature) < 0, S2N_ERR_BAD_MESSAGE);
+    S2N_ERROR_IF(s2n_pkey_verify(&conn->secure.server_public_key, signature_hash, &signature) < 0, S2N_ERR_BAD_MESSAGE);
 
     /* We don't need the key any more, so free it */
     GUARD(s2n_pkey_free(&conn->secure.server_public_key));
     return 0;
 }
 
-static int s2n_dhe_server_key_recv(struct s2n_connection *conn)
+int s2n_ecdhe_server_key_recv(struct s2n_connection *conn, struct s2n_blob *data_to_sign)
+{
+    struct s2n_stuffer *in = &conn->handshake.io;
+
+    /* Read server ECDH params and calculate their hash */
+    GUARD(s2n_ecc_read_ecc_params(&conn->secure.server_ecc_params, in, data_to_sign));
+    return 0;
+}
+
+int s2n_dhe_server_key_recv(struct s2n_connection *conn, struct s2n_blob *data_to_sign)
```

Comment:
The name of the functions does no longer represent their implementation. What it should have being called is s2n_dhe_server_key_recv_get_params_data, think on something smaller. Because now all the complexity of the key_recv is in the general  function above.

Please also note that data_to_sign is "weird" here because the signing does not happen here. Instead it should be something like dhparams....